File: programmers/pretest6.java
package programmers;

public class pretest6{

public static void main(String[] args){

    int[] sticker = {14,6,5,11,3,9,2,10}; //36
    System.out.println(solution(sticker));
}
public static int solution(int[] sticker){
    
    int answer =0;

    int len = sticker.length;
    if(len==1) return sticker[0];

    int[] dp1 = new int[100001];
    int[] dp2 = new int[100001];

    //1번 스티커 뜯을 경우
    dp1[0] = sticker[0];
    dp1[1] = dp1[0];

    for(int i=2;i<len-1;i++){
        dp1[i] = Math.max(dp1[i-1],dp1[i-2] + sticker[i]);
    }

    //1번 스티커 안뜯을 경우
    dp2[0] = 0;
    dp2[1] = sticker[1];

    for(int i=2;i<len;i++){
        dp2[i] = Math.max(dp2[i-1],dp2[i-2] + sticker[i]);
    }

    return Math.max(dp1[len-2],dp2[len-1]);
    }
}

File: programmers/pretest6.cpp
#include<vector>
using namespace std;

int dp1[100001]; //첫번째 스티커를 뜯은 경우
int dp2[100001]; //첫번째 스티커를 뜯지 않은 경우

int pretest6(vector<int> sticker)
{
    int n = sticker.size();
//스티커가 한장이라면
    if(n == 1)
        return sticker[0];

//스티커가 적어도 2장 이상인경우

    //첫 번째 스티커를 뜯은 경우
    //첫번째 인덱스와 두번째 인덱스를 초기화 시켜준다.
    dp1[0] = sticker[0];
    dp1[1] = dp1[0];

    //첫번째 스티커 반드시 뜯은 경우 n-1번째 스티커를 뜯으면 안되기 때문에 n-2까지 for문 돈다.(dp가 0부터 시작해서 n-1이 마지막 숫자임)
    for(int i=2;i<n-1;i++){
        dp1[i] = max(dp1[i-1],dp1[i-2] + sticker[i]);
        //dp1[n-2]에 답이 들어있게 됩니다.첫번쨰 스티커부터 마지막스티커까지 적절히 잘뜯었을때 들어있는 최대 정수
    
    
    //첫번째 스티커를 뜯지 않은 경우
    dp2[0] = 0;
    dp2[1] = dp2[0];
    for(int i=2;i<n;i++){
        dp2[i] = max(dp2[i-1],dp2[i-2] + sticker[i]);
        //dp2[n-1]에 답이 들어있게 됩니다.
    }
    }
   // int answer = 36;
    return max(dp1[n-2],dp2[n-1]);
}